spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/actions/SparkShufflingDataRewriter.java

```diff
@@ -59,7 +61,24 @@ abstract class SparkShufflingDataRewriter extends SparkSizeBasedDataRewriter {
 
   public static final double COMPRESSION_FACTOR_DEFAULT = 1.0;
 
+  /**
+   * The number of shuffle partitions to use for each output file. By default, this file rewriter
+   * assumes each shuffle partition would become a separate output file. Attempting to generate
+   * large output files of 512 MB and more may strain the memory resources of the cluster as such
+   * rewrites would require lots of Spark memory. This parameter can be used to further divide up
+   * the data which will end up in a single file. For example, if the target file size is 2 GB, but
+   * the cluster can only handle shuffles of 512 MB, this parameter could be set to 4. Iceberg will
+   * use a custom coalesce operation to stitch these sorted partitions back together into a single
+   * sorted file.
+   *
+   * <p>Note using this parameter requires enabling Iceberg Spark session extensions.
+   */
+  public static final String SHUFFLE_PARTITIONS_PER_FILE = "shuffle-partitions-per-file";
```

Comment:
Not to block this change, but did we consider having shuffle-threshold?  Ie, if we have some partition with 2G but others that are way less than 512MB, no need to shuffle the ones that are less?

Reply:
You mean like switching to a local sort if the size of the data to compact is small?

Reply:
I was just wondering the use case, where we set shuffle-partitions-per-file to 4, because we want 4GB files but can only shuffle 512mb.  However, consider an Iceberg partition (rewrite group) that has only 512MB files.  Will we still shuffle to four partitions in this case and coalesce at end, unnecessarily?  I may be missing something.

Reply:
It should be still fine to apply this optimization as there is no extra cost. I achieved best results with 128 MB shuffle blocks so it should be fairly safe to assume the operation would complete fine.

Reply:
I see, but would there be issues in contending for pods.  Also wouldn't it make more sense to have 128MB as a conf  (shuffle-threshold), otherwise its always a bit dynamic depending on the max partition size?  Not sure if there are other issues with this approach.

Reply:
To be honest, I have never seen issues with this approach in any of our prod jobs in the last few years. Not applying this split if the size of the job is less than 128MB could be a valid step but it would require quite a bit of changes to pass more info around. I'd probably skip it for now until we experience any issues.

Reply:
Sure, we can do it later then if there's a need